### PR TITLE
fix: skip checking rewrites with an empty has array in isInterceptionRouteRewrite

### LINKED ISSUE #63871: HookWebpackError in isInterceptionRouteRewrite() when rewrites has an empty `has` array
### Link to the code that reproduces this issue

[CodeSandbox](https://codesandbox.io/p/devbox/inspiring-pasteur-rrjp64?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluboxdgr0008356ijzb29007%2522%252C%2522sizes%2522%253A%255B38.693830357199175%252C61.306169642800825%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluboxdgr0003356iy4vnfzq1%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluboxdgr0006356ikvxtlm7e%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluboxdgr0007356ie0vrz84w%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B79.18835351713476%252C20.811646482865243%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluboxdgr0003356iy4vnfzq1%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluboxdgr0002356i7n60fo6s%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fnext.config.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cluboxdgr0003356iy4vnfzq1%2522%252C%2522activeTabId%2522%253A%2522cluboxdgr0002356i7n60fo6s%2522%257D%252C%2522cluboxdgr0007356ie0vrz84w%2522%253A%257B%2522id%2522%253A%2522cluboxdgr0007356ie0vrz84w%2522%252C%2522activeTabId%2522%253A%2522cluczap8d02vr356ja9eyui26%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522cluczap8d02vr356ja9eyui26%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cluboxdgr0006356ikvxtlm7e%2522%253A%257B%2522id%2522%253A%2522cluboxdgr0006356ikvxtlm7e%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522npm%2520run%2520build%2522%252C%2522id%2522%253A%2522clubozeoj0028356jlbxi4ctd%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522id%2522%253A%2522clubp1gu800ae356jspmqlfbz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clubp1gxb00sudgglf67wdhif%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clubozeoj0028356jlbxi4ctd%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A20.989583333333343%257D)

### To Reproduce

From an empty Next.js project, adding an empty array for a `has` rewrite option that's `beforeFiles` triggers a HookWebpackError when building.

### Current vs. Expected behavior

No errors occur when building the Next.js project, even if the rewrite `has` array is empty, just like before the v14.2.0 canaries

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #16~22.04.1-Ubuntu SMP Wed Mar 13 18:54:49 UTC 2024
  Available memory (MB): 15799
  Available CPU cores: 4
Binaries:
  Node: 20.11.0
  npm: 10.2.4
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.0-canary.47 // Latest available version is detected (14.2.0-canary.47).
  eslint-config-next: 13.4.8
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.4.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Routing (next/router, next/navigation, next/link)

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

This is due to `isInterceptionRouteRewrite` from #61794 assumes that if the route `has` is truthy, then there is at least one `has` item.

### PR DESCRIPTION
**Pull request by evelynhathaway**
### What?

Adds safe traversal to rewrite `has` items in a certain new (`isInterceptionRouteRewrite()` added in the v14.2.0 canaries) check to the rewrite routes.

### Why?

The new check assumed that all `has` arrays had at least one item, when previously Next.js accepted rewrites with empty `has`. Adding safe traversal doesn't negatively impact the check, as the check only needs rewrites with the first item.

Fixes #63871

### CODE REVIEW

**Review by ijjk — COMMENTED** [review]
Seems like a good check to add, could we also add a test case for this by having a rewrite with empty `has` potentially here https://github.com/vercel/next.js/blob/ff1db77f1bd1298c68debace25ff7062bfcefcda/test/e2e/app-dir/app/next.config.js#L9

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| buildDuration | 14.5s | 14.7s | ⚠️ +115ms |
| buildDurationCached | 9.1s | 6.9s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +590 B |
| nextStartRea..uration (ms) | 411ms | 413ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.4 kB | 31.4 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 241 B | N/A |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99 kB | 99 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| index.html gzip | 530 B | 529 B | N/A |
| link.html gzip | 542 B | 542 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 542 B | 542 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.3 kB | 95.3 kB | N/A |
| page.js gzip | 3.04 kB | 3.04 kB | ✓ |
| Overall change | 3.04 kB | 3.04 kB | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 97 kB | 97 kB | ✓ |
| app-page-tur..prod.js gzip | 98.8 kB | 98.8 kB | ✓ |
| app-page-tur..prod.js gzip | 93 kB | 93 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.5 kB | 91.5 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.1 kB | 21.1 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.4 kB | 22.4 kB | ✓ |
| server.runti..prod.js gzip | 51 kB | 51 kB | N/A |
| Overall change | 894 kB | 894 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | evelynhathaway/next.js patch-2 | Change |
| - | - | - | - |
| 0.pack gzip | 1.58 MB | 1.58 MB | N/A |
| index.pack gzip | 107 kB | 106 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 5abfb4fd8aaf082eb319e78b64eb0d75234d988b

File: test/e2e/app-dir/app/next.config.js [hunk 1]
```diff
@@ -8,6 +8,14 @@ module.exports = {
   // output: 'standalone',
   rewrites: async () => {
     return {
+      beforeFiles: [
+        {
+          source: '/before-files-rewrite-with-empty-arrays',
+          destination: '/',
+          has: [],
```
Comment (evelynhathaway) [inline test/e2e/app-dir/app/next.config.js:15]: `beforeFiles` with an empty `has` array is all that's needed to make the tests fail without the fix. I added the `missing` array, and did the same for `afterFiles` and `fallback` for completeness.

**Review by evelynhathaway — COMMENTED** [review]
Thank you @ijjk for the very helpful callout for how to test this.

I added the rewrites and successfully ran the tests (successful only with the fix).

**Review by ijjk — APPROVED** [review]
Thanks for the PR!

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 97853eaf84e579a59f453b1b91d7769b8c8c4a93

`pnpm test test/integration/edge-runtime-module-errors/test/module-imports.test.js`
- Edge runtime code with imports > Edge API statically importing node.js module > throws unsupported module error in dev at runtime and highlights the faulty line

<details>
<summary>Expand output</summary>

  ● Edge runtime code with imports › Edge API statically importing node.js module › throws unsupported module error in dev at runtime and highlights the faulty line

    expect(received).toBe(expected) // Object.is equality

    Expected: 500
    Received: 404

      86 |       context.app = await launchApp(context.appDir, context.appPort, appOption)
      87 |       const res = await fetchViaHTTP(context.appPort, url)
    > 88 |       expect(res.status).toBe(500)
         |                          ^
      89 |
      90 |       const text = await res.text()
      91 |       await check(async () => {

      at Object.toBe (integration/edge-runtime-module-errors/test/module-imports.test.js:88:26)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 97853eaf84e579a59f453b1b91d7769b8c8c4a93

`TURBOPACK=1 pnpm test-start test/e2e/prerender.test.ts` (turbopack)
- Prerender > should on-demand revalidate for fallback: blocking with onlyGenerated if generated

<details>
<summary>Expand output</summary>

  ● Prerender › should on-demand revalidate for fallback: blocking with onlyGenerated if generated

    expect(received).toMatch(expected)

    Expected pattern: /(HIT|STALE)/
    Received string:  "MISS"

      2297 |
      2298 |         expect(initialTime).toBe($2('#time').text())
    > 2299 |         expect(res2.headers.get('x-nextjs-cache')).toMatch(/(HIT|STALE)/)
           |                                                    ^
      2300 |
      2301 |         const res3 = await fetchViaHTTP(
      2302 |           next.url,

      at Object.toMatch (e2e/prerender.test.ts:2299:52)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 97853eaf84e579a59f453b1b91d7769b8c8c4a93

`pnpm test test/integration/edge-runtime-module-errors/test/module-imports.test.js`
- Edge runtime code with imports > Edge API statically importing node.js module > throws unsupported module error in dev at runtime and highlights the faulty line

<details>
<summary>Expand output</summary>

  ● Edge runtime code with imports › Edge API statically importing node.js module › throws unsupported module error in dev at runtime and highlights the faulty line

    expect(received).toBe(expected) // Object.is equality

    Expected: 500
    Received: 404

      86 |       context.app = await launchApp(context.appDir, context.appPort, appOption)
      87 |       const res = await fetchViaHTTP(context.appPort, url)
    > 88 |       expect(res.status).toBe(500)
         |                          ^
      89 |
      90 |       const text = await res.text()
      91 |       await check(async () => {

      at Object.toBe (integration/edge-runtime-module-errors/test/module-imports.test.js:88:26)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).
